### ISSUE TITLE
Error dealing with the splat operator and numeric array keys

### ISSUE THREAD

**Issue by mpdude** [vimeo/psalm#8190]
Please consider:

```php
<?php

/**
 * @return array{ 0: string, 1?: array }
 */
function foo(): array
{
	return ['test'];
}

function bar(string $first, array $second = []): void
{
	print_r($first);
    print_r($second);
}

$x = foo();
bar(...$x);
```

This gives:

> Psalm output (using commit 7dff408): 
> INFO: [MixedArgument](https://psalm.dev/030) - 18:8 - Argument 1 of bar cannot be mixed, expecting string
> ERROR: [TooFewArguments](https://psalm.dev/025) - 18:1 - Too few arguments for bar - expecting first to be passed

https://psalm.dev/r/f3957ce2e7

**Comment by AndrolGenhald** [comment]
Well, `TooFewArguments` is a false-positive, but this isn't really possible to handle correctly now since array-shapes [allow extra keys](https://psalm.dev/r/85f9c121d2). I think we'll just have to wait till we have #5299 done and then basically redo this check from the ground up.
